### PR TITLE
[BEAM-4654] Treat timers as PCollections within proto representation in the Java SDK.

### PR DESCRIPTION
**Pull request by lukecwik**
I abstracted away PTransform translation so that timers can be translated to PCollections since the existing translation mechanism required could only modify the FunctionSpec.
Note that I took this approach because changing the AppliedPTransform implementation to treat timers as an additional input and output would require updating all the non portable runners which did not seem worth the effort.
Once portability lands, we can go back and clean up several user and Runner facing interfaces which contain both user and Runner specific bindings.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by lukecwik** [comment]
CC: @charlesccychen

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
R: @angoenka @axelmagn

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java [hunk 1]
```diff
@@ -88,39 +91,72 @@
...
+              .build());
+
+      String mainInputId = getMainInputId(builder, payload);
+      PCollection<KV<?, ?>> mainInput =
+          (PCollection) appliedPTransform.getInputs().get(new TupleTag(mainInputId));
+
+      // https://s.apache.org/beam-portability-timers
+      // Add a PCollection and coder for each timer. Also treat them as inputs and outputs.
+      for (String localTimerName : payload.getTimerSpecsMap().keySet()) {
+        PCollection<?> timerPCollection =
```
Comment (angoenka) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java]: How is TimerSpec encoded in this PCollectoon?

Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDoTranslation.java]: The PCollection has a timer coder on it as defined in https://github.com/apache/beam/pull/5789 and an example implementation in https://github.com/apache/beam/pull/5794
For all our current timer PCollections, it is always KV<K, Timer<Payload>>. Java doesn't expose the payload so it is always `null`.

**Comment by angoenka** [comment]
Thanks!

**Pull request merged**